session/txn.go

```diff
@@ -516,6 +516,11 @@ func (s *session) StmtCommit(memTracker *memory.Tracker) error {
 		// in memTracker to avoid double-counting. If it's not batch mode, this
 		// work has no effect because that no more data will be appended into
 		// s.txn.
+
+		if r := recover(); r != nil {
+			err := errors.Errorf("%v", r)
+			s.txn.doNotCommit = err
+		}
```

Comment:
@bobotu @XuHuaiyu
The problem remains.
```
drop table if exists t;
create table t(a int);
begin;
insert into t values(1), (-1), (0);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
tidb> insert /*+ memory_quota(1 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(2 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(14 MB) */ into t select * from t;
Query OK, 393216 rows affected (1.60 sec)
Records: 393216  Duplicates: 0  Warnings: 0
tidb> insert /*+ memory_quota(15 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> commit;
Query OK, 0 rows affected (1.04 sec)

mysql> select count(*) from t;
+----------+
| count(*) |
+----------+
|   786432 |
+----------+
1 row in set (0.07 sec)
```

Reply:
You need to change the definition of `StmtCommit` to 

```go
func (s *session) StmtCommit(memTracker *memory.Tracker) (err error)
```

Then add change the code you added to

```go
if r := recover(); r != nil {
	err = errors.Errorf("%v", r)
	s.txn.doNotCommit = err
}
```

Reply:
We need to return an error to the caller.